applicationset/services/scm_provider/azure_devops.go

```diff
@@ -0,0 +1,157 @@
+
+func NewAzureDevopsProvider(ctx context.Context, accessToken string, org string, project string) (*AzureDevopsProvider, error) {
+	if accessToken == "" {
+		return nil, fmt.Errorf("No access token provided")
+	}
+
+	baseUrl, err := getBaseUrl()
+	if err != nil {
+		return nil, err
+	}
+	url := fmt.Sprintf("%s/%s", *baseUrl, org)
+	connection := azuredevops.NewPatConnection(url, accessToken)
+	return &AzureDevopsProvider{organization: org, teamProject: project, accessToken: accessToken, clientFactory: &devopsFactoryImpl{connection: connection}}, nil
+}
+
+func (g *AzureDevopsProvider) ListRepos(ctx context.Context, cloneProtocol string) ([]*Repository, error) {
+	gitClient, err := g.clientFactory.GetClient(ctx)
+	if err != nil {
+		return nil, err
+	}
+	getRepoArgs := azureGit.GetRepositoriesArgs{Project: &g.teamProject}
+	azureRepos, err := gitClient.GetRepositories(ctx, getRepoArgs)
+
+	if err != nil {
+		return nil, err
+	}
+	repos := []*Repository{}
+	for _, azureRepo := range *azureRepos {
+		if azureRepo.Name == nil || azureRepo.DefaultBranch == nil || azureRepo.RemoteUrl == nil || azureRepo.Id == nil {
+			continue
+		}
+		repos = append(repos, &Repository{
+			Organization: g.organization,
+			Repository:   *azureRepo.Name,
+			URL:          *azureRepo.RemoteUrl,
+			Branch:       *azureRepo.DefaultBranch,
+			Labels:       []string{},
+			RepositoryId: *azureRepo.Id,
+		})
+	}
+
+	return repos, nil
+}
+
+func (g *AzureDevopsProvider) RepoHasPath(ctx context.Context, repo *Repository, path string) (bool, error) {
+	gitClient, err := g.clientFactory.GetClient(ctx)
+	if err != nil {
+		return false, err
+	}
+
+	idAsString := fmt.Sprintf("%v", repo.RepositoryId)
+	branchName := repo.Branch
+	getItemArgs := azureGit.GetItemArgs{RepositoryId: &idAsString, Project: &g.teamProject, Path: &path, VersionDescriptor: &azureGit.GitVersionDescriptor{Version: &branchName}}
+	_, err = gitClient.GetItem(ctx, getItemArgs)
+
+	if err != nil {
+		if wrappedError, isWrappedError := err.(azuredevops.WrappedError); isWrappedError {
+			if *wrappedError.TypeKey == "GitItemNotFoundException" {
+				return false, nil
+			}
+		}
+		return false, err
+	}
+
+	return true, nil
+}
+
+func (g *AzureDevopsProvider) GetBranches(ctx context.Context, repo *Repository) ([]*Repository, error) {
+	gitClient, err := g.clientFactory.GetClient(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	getBranchesRequest := azureGit.GetBranchesArgs{RepositoryId: &repo.Repository, Project: &g.teamProject}
+	branches, err := gitClient.GetBranches(ctx, getBranchesRequest)
+	if err != nil {
+		return []*Repository{}, nil //Repo might be locked/unavailable, branch authz; all sorts of reasons why this would fail. Just return empty result.
+	}
+	repos := []*Repository{}
+	for _, azureBranch := range *branches {
+		repos = append(repos, &Repository{
+			Branch:       *azureBranch.Name,
+			SHA:          shortCommitSHA(*azureBranch.Commit.CommitId),
```

Comment:
Is it important to use the short hash instead of the full hash?

Reply:
Initially, I used the commit hash as a way of creating unique release names. But with the addition of the `branchNormalized` parameter I've changed the code to use the full commit hash.